### PR TITLE
attempt to optimize prompts

### PR DESCRIPTION
**Pull request by jaykchen**
tried to optimize prompts by:
1) observed that agent goes off track often, so
- from beginning, set the agent's capabilities scope straight
- revise other prompts trying to help agent to focus on intended task

2) feel that some prompts can be shortened a bit to save tokens
- asking gpt4 to optimize prompt to use less tokens
- test to check whether original result is retained
- iteratively optimize

### CODE REVIEW

File: src/auto/agents/manager.rs [hunk 1]
```diff
@@ -40,28 +56,47 @@ r#"
 Personality: {personality}
 
 You will be given one task.
-Your goal is to split that task up into smaller tasks.
-Your goal is to delegate those subtasks, one at a time.
+Your goal is to understand the task, split it into smaller tasks when necessary.
```
Comment (jaykchen) [inline src/auto/agents/manager.rs:59]: tell the agent to stay within capabilities scope
ask the agent to go straight/quickly to commands

File: src/auto/agents/manager.rs [hunk 2]
```diff
@@ -85,30 +120,36 @@ Reply in this format:
 }}
 ```
 
-Ensure your delegated task is explained in detail.
+Ensure your delegated task is explained in detail but not overly verbose.
```
Comment (jaykchen) [inline src/auto/agents/manager.rs:123]: remind llm not to talk more than needed

File: src/auto/classify.rs [hunk 3]
```diff
@@ -1,52 +1,57 @@
...
-}}
-```"#)));
-
-    context.agents.fast.llm.message_history.push(Message::User(format!(
-        "Request to Classify: {task}"
-    )));
+"thoughts on how to classify it": "<thoughts_here>",
+"message classification": "<classification_here>"
+}}"#,
+        )));
```
Comment (jaykchen) [inline src/auto/classify.rs:45]: limited test shows that this shortened version is enough to make llm produce consistent output as original version

File: src/auto/agents/manager.rs [hunk 4]
```diff
@@ -40,28 +56,47 @@ r#"
...
+When parsing user instructions, please prioritize the command the user instructed you to use to do the job, when splitting the tasks into subtasks please consider your current capabilities pool: 
+- Browse (commands: browse_url)
+- Google (commands: google_search)
+- File System (commands: file_write, file_append, file_list, file_read)
+- Wolfram (commands: wolfram)
+- ChatGPT (commands: ask_chatgpt, reset_chatgpt)
+- NewsAPI (commands: news_search)
+- Wikipedia (commands: wikipedia_search, wikipedia_browse)
+
+Your goal is to map the subtasks to your capabilities, delegate those subtasks, one at a time.
```
Comment (jaykchen) [inline src/auto/agents/manager.rs:69]: tell the agent to stay within capabilities scope
ask the agent to go straight/quickly to commands

File: src/auto/responses.rs [hunk 5]
```diff
@@ -2,55 +2,66 @@ use std::error::Error;
...
-    "response": "..."
+    "response": "<Your Response (Source)>"
 }}
-
-Respond in that exact JSON format exactly.
-
-Provide a response as an assistant to the initial request in the above format.
-Make sure you include where you got the information from in your response, in parantheses."#)
+"#
+    )
```
Comment (jaykchen) [inline src/auto/responses.rs:46]: limited test shows that this shortened version is enough to make llm produce consistent output as original version

File: src/auto/classify.rs [hunk 6]
```diff
@@ -1,52 +1,57 @@
...
-Respond in this format:
-
-```json
+    context
+        .agents
+        .fast
+        .llm
+        .prompt
+        .push(Message::Assistant(String::from(
+            "Respond with either 'conversational' for a chat message or 'task' for a request.",
```
Comment (jaykchen) [inline src/auto/classify.rs:31]: shorter wording

File: src/auto/classify.rs [hunk 7]
```diff
@@ -1,52 +1,57 @@
...
-    "message classification": "..."
-}}
-```"#)));
-
-    context.agents.fast.llm.message_history.push(Message::User(format!(
-        "Request to Classify: {task}"
-    )));
+"thoughts on how to classify it": "<thoughts_here>",
+"message classification": "<classification_here>"
+}}"#,
```
Comment (jaykchen) [inline src/auto/classify.rs:44]: functional equivalent prompt wording

File: src/auto/mod.rs [hunk 8]
```diff
@@ -44,7 +44,7 @@ pub fn run_assistant_auto(program: &mut ProgramInfo, messages: &[Message], reque
 
     let mut new_messages = messages.to_vec();
     new_messages.push(Message::User(format!(
-r#"Summarize the conversation."#)));
+r#"Summarize the conversation. Focus on the user's intension and what has been achieved so far."#)));
```
Comment (jaykchen) [inline src/auto/mod.rs:47]: intent, intent, intent

File: src/auto/agents/manager.rs (hunk 4 above)
Comment (Cormanz) [reply src/auto/agents/manager.rs:69]: Adding the capabilities directly has been tried before, but lead to the problem of the manager over specifying what tools to use, and hurting the organization of the Manager and Employee.

In addition, the way you hardcoded it into the prompt subverts the way that toggling plugins on and off works.

**Comment by Cormanz** [comment]
I really appreciate all of the work in this PR, but given that #28 completely reworks this entire system and is close to being merged, I'm not sure it's worth it to merge this once it's ready.

**Comment by jaykchen** [comment]
understood, will check back later

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>
